### PR TITLE
Implement changes to target Android SDK 30

### PR DESCRIPTION
**Pull request by ewpatton**
This change makes use of the new helper blocks to provide robust access to different parts of the Android file system under the latest constraints imposed by Google. Due to the complexity of this change, it is broken into some smaller commits to make it easier to review. Each change in brief:

1. Revert closure-library change (also part of #2511)- Not strictly needed in this change and should be rebased out once the other PR is merged
2. Fix helper blocks not working in emulator- For backwards compatibility, the helper blocks send a macro to the companion that is used to protect against accessing the undefined classes on older companion versions. However, this macro was only sent in the Wifi codepath and not the emulator/USB code path, which prevented the logic from working in the emulator
3. Fix helper codegen for non-hygenic macros- The helper blocks use Kawa's colon notation to reference the static enum fields. However, Kawa's reader seems to process this differently from within a non-hygenic macro (e.g., foreach). Therefore, helper blocks work fine outside of certain block constructs but cause read errors within others. This change switches to using the more verbose `(static-field class field)` lookup, which behaves consistently when used within a macro.
4. Refactor project settings to use a builder pattern- This makes it more convenient to add/manage project settings. It also allows for the fact that settings could be semantically the same even if the serialized content ends up in a different line order.
5. Refactor Synchronizer out of MediaUtil- This is a stepping stone for the following change...
6. Implement limited continuations for the blocks language- This change adds support for allowing components to take a continuation parameter for continuation passing style. Because we do not implement CPS in the code generator, component functions at this time MUST block the UI thread and invoke the continuation on the UI thread once complete. This restriction can be relaxed once full CPS is implemented. This is used by the new method blocks in the File component introduced in...
7. Implement scoped storage in accordance with SDK 30 requirements- Adds a new FileScope helper block for specifying what scope a file should be put in. The default is App specific storage (available on Froyo and up). THIS IS A BREAKING CHANGE. Existing apps that need access to files outside of the app-specific storage should switch to using the Legacy scope, which provides the old behavior. Existing curriculum and tutorials should not be affected by this change if followed from beginning to end within a fresh project.
8. Update to SDK 30- Updates the android.jar and various other flags/constants so that compiled apps target SDK 30.

TODOs:

- [ ] More testing of the shared storage scope
- [ ] Add documentation for the File component to discuss the different scopes and how they work
- [ ] Create some example apps showing off the new File functionality
- [ ] iOS support for the new properties and methods

### CODE REVIEW

**Comment by ewpatton** [comment]
I've only tested this on SDK 7 emulator right now. There's still more testing to be done on different SDK levels. I'm putting this here mainly to get feedback on the design and implementation. It should be 100% backward compatible with the existing logic (i.e., everything that was valid before will also be valid here, at least until App Inventor targets Android 11). This adds the further option beyond #2268 to let apps uses the app-specific directory on external storage on earlier versions of Android to assist in migrating data. It will become more powerful with the migrate function @jisqyv has proposed.

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidSettingsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    wrapper.put(PROJECT_YOUNG_ANDROID_SETTINGS, object);
+    return wrapper.toString();
+  }
+
+  /**
+   * Convert the internal settings into a Java Properties object.
+   *
+   * @return a properties file contents containing the settings
+   */
+  public String toProperties() {
```
Comment (pavi2410) [inline appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidSettingsBuilder.java:235]: Maybe I am overlooking something, but I don't see `aname` prop here

File: appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java [hunk 2]
```diff
@@ -585,4 +820,261 @@ public int getErrorMessageNumber() {
...
+
+  private static Uri getContentUriForPath(String path) {
+    if ("DCIM".equals(path) || "Pictures".equals(path) || "Screenshots".equals(path)) {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        return MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL);
+      }
+      return MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
+    } else if ("Videos".equals(path) || "Movies".equals(path)) {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        return MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL);
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java:1066]: This should return `MediaStore.Video`, shouldn't it?

File: appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java [hunk 3]
```diff
@@ -585,4 +820,261 @@ public int getErrorMessageNumber() {
...
+      }
+      return MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
+    } else if ("Audio".equals(path) || "Music".equals(path)) {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        return MediaStore.Video.Media.getContentUri(MediaStore.VOLUME_EXTERNAL);
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java:1071]: should be `MediaStore.Audio` here

File: appinventor/components/src/com/google/appinventor/components/runtime/File.java [hunk 4]
```diff
@@ -16,64 +21,118 @@
...
  * Non-visible component for storing and retrieving files. Use this component to write or read files
- * on the device. The default behavior is to write files to the private data directory associated
- * with the app. The Companion writes files to `/sdcard/AppInventor/data` for easy debugging. If
- * the file path starts with a slash (`/`), then the file is created relative to `/sdcard`.
- * For example, writing a file to `/myFile.txt` will write the file in `/sdcard/myFile.txt`.
+ * on the device. File names can take one of three forms:
+ *
+ * - Private files have no leading `/` and are written to app private storage (e.g., "file.txt")
+ * - External files have a single leading `/` and are written to public storage (e.g., "/file.txt")
+ * - Bundled app assets have two leading `//` and can only be read (e.g., "//file.txt")
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/File.java:61]: I am not getting this. If the `File Scope` is set to `Asset`, then does the file name require to be prefixed by `//`?

File: appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java (hunk 3 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java:1071]: Yes, you're right. I'll fixed this in the next iteration.

File: appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java (hunk 2 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/runtime/util/FileUtil.java:1066]: Yes.

File: appinventor/components/src/com/google/appinventor/components/runtime/File.java (hunk 4 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/runtime/File.java:61]: No, it does not require the `//`. In any case, if you use the `//` prefix then the scope will be switched to `Asset`. I will think about how to refine this to be more clear.

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidSettingsBuilder.java (hunk 1 above)
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidSettingsBuilder.java:235]: Yes, I'm not sure how I missed that key when I was translating the code over.

**Review by jisqyv — APPROVED** [review]
LGTM

**Pull request merged**